[Kelvin Fong Koi Mun] iP

nusE0726844 opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke helps you manage your tasks for you. It is,
- text-based
- **FREE!!** :star_struck:

### How to get it
All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2223S2/ip)
2. add your tasks

And you're done!

### Features:
- [x] Keeping track of list of tasks
- [x] Keeping track of completion status of task
- [ ] Reminders (coming soon)

Using Duke is simple. Here is the `main` method:
``` ruby 
    public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

nusE0726844 committed:
Level 1: Greet, Echo, Exit

nusE0726844 committed:
Level 2. Add, List

nusE0726844 committed:
Level 3. Mark as Done

nusE0726844 committed:
Level 3. Mark as Done

nusE0726844 committed:
Level 4. ToDos, Events, Deadlines

nusE0726844 committed:
Automated Text UI Testing. Test using the I/O redirection technique

nusE0726844 committed:
no message

nusE0726844 committed:
no message

nusE0726844 committed:
Level 5. Handle Errors

nusE0726844 committed:
Level 6. Delete

nusE0726844 committed:
Uses Enumerations

nusE0726844 committed:
Fixed event

nusE0726844 committed:
Add greet, echo, and exit

Duke Class: add
* greet(): print user greeting
* echo(): print specified message
* exit(): print exiting message

nusE0726844 committed:
Add add and list

Duke Class: add
* add(): print user greeting
* list(): print specified message
* tasks: store input as task

Duke Class: change
* decision tree if statement to switch expression

nusE0726844 committed:
Add mark, unmark, and Task

Task Class: contain information on the task

Duke Class: add
* mark(): mark specified task as done
* unmark(): change status of specified task back
            to not done
* indent(): return indentation with spaces

Duke Class: change
* header comment: capitalise header comment
                  parameter starting word
* add(): change order of parameters
* echo(): overload echo to accept array of messages
          to be printed to console at the same time

nusE0726844 committed:
Add ToDo, Deadline, and Event

Add
* ToDo Class: contain only information on the task
* Deadline Class: contain information and the deadline
                  of the task
* Event Class: contain information and the duration
               on the task
* add(): overload method to accept arraylist and
         task object

nusE0726844 committed:
Add DukeException

ToDo Class: add
* Error message for empty description

Deadline Class: add
* Error message for empty description
* Error message for empty deadline

Event Class: add
* Error message for empty description
* Error message for empty start time
* Error message for empty end time

Mark/Unmark Class: add
* Error message for empty index
* Error message for index out of bound
* Error message for non integer index

Add generate method to handle exceptions for
* ToDo Class
* Deadline Class
* Event Class

nusE0726844 committed:
Add delete

Duke Class: add
* delete(): remove task from the list of tasks

DukeException Class: add error messages for delete
* Error message for empty index
* Error message for index out of bound
* Error message for non integer index

nusE0726844 committed:
Add enumerations

Add enums
* Command: for valid commands
* ErrorType: for valid error types

nusE0726844 committed:
Add save and load

Duke Class: Add
* saveTask(): save changes made to list of tasks
              to hard disk
* loadTask(): load list of task on startup
* getTaskFile(): get path to data of saved tasks

Add load(): use saved data to create task
Add load() to:
* Task Class
* Todo Class
* Deadline Class
* Event Class

nusE0726844 committed:
Change format of stored datetime

Deadline Class:
* change by String type to LocalDate type
* add behaviour to check if reached deadline

Event Class: change
* from: String type to LocalDate type
* to: String type to LocalDate type

Event Class: add behaviour to
* check if event over
* check if end time of event is before start time
  of event

nusE0726844 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/DukeException.java
#	src/main/java/Event.java

slackernoob commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Creates a deadline object
+     *
+     * @param description The description of the deadline
+     * @param by Deadline time of the deadline
```
Perhaps end param description with a punctuation, e.g. fullstop.

slackernoob commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,315 @@
+    private static final String MSG_UNMARK = "OK, I've marked this task as not done yet:";
+
+    /** Message to be printed after adding task to list of tasks */
+    private static final String[] MSG_ADD = {"Got it. I've added this task:", "Now you have ", " tasks in the list."};
```
Perhaps split the line into multiple lines to keep within the 120 hard limit for characters.

slackernoob commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,315 @@
+        ArrayList<Task> tasks = new ArrayList<>();
+
+        // Decision loop
+        boolean isContinue_decisionLoop = true;
```
Maybe name the variable in camel case? Since it is not a method used in testing.

slackernoob commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
Perhaps indent the comment relative to its position in the code.

slackernoob commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,161 @@
+public class DukeException extends Exception {
+
+    private static final String MSG_ERR_PREFIX = "☹ OOPS!!!";
+    private static final String MSG_ERR_SUFFIX = ":-(";
```
Liked the usage of constants to keep the code nice and tidy. Looks good!

slackernoob reviewed (commented):
Overall, LGTM!. I found your code easy to read for the most part except a few places
where lines exceeded the hard character limit and was slightly too long.  However I liked the usage of constants, and the strict adherence to screaming snake case, making the code look neat and tidy. Great job!

nusE0726844 committed:
Make the code more OOP

Add Class
* Ui: deals with interactions with the user
* Storage: deals with saving and loading tasks
           from the file
* Parser: deals with making sense of the user command
* TaskList: contains the list of tasks

* Command: an abstract class to be inherited from
* addCommand: handles adding tasks to tasks
* deleteCommand: handles removing tasks to tasks
* markCommand: handles marking tasks as done
* unmarkCommand: handles marking tasks as not done
* listCommand: handles showing list of task
* exitCommand: handles exiting Duke

Moved method:
* add() to addCommand
* delete() to deleteCommand
* mark() to markCommand
* mark() to unmarkCommand
* list() to listCommand

nusE0726844 committed:
Divide classes into packages

Move all classes to duke paskage

Move classes to duke.command package
* Command
* addCommand
* deleteCommand
* markCommand
* unmarkCommand
* listCommand
* exitCommand

Move classes to duke.task package
* Task
* Todo
* Deadline
* Event

Rename classes
* addCommand: AddCommand
* deleteCommand: DeleteCommand
* markCommand: MarkCommand
* unmarkCommand: UnmarkCommand
* listCommand: ListCommand
* exitCommand: ExitCommand

nusE0726844 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

nusE0726844 committed:
Set up gradle to build and run Duke

nusE0726844 committed:
Add JUnit tests to test the behavior of the code

Add Class:
* DukeTest: contains dummyTest
* TaskTest: tests Task methods
* TodoTest:  tests Todo methods
* DeadlineTest:  tests Deadline methods
* EventTest:  tests Event methods

nusE0726844 committed:
Package the App as a JAR file

JAR file available at tag v0.1

nusE0726844 committed:
Add JavaDoc comments to the code

Add header comments to:
duke
* Duke Class
* DukeException Class
* Parser Class
* Storage Class
* TasksList Class
* Ui Class

duke.task
* Task Class
* Todo Class
* Deadline Class
* Event Class

duke.command
* Command Class
* ExitCommand Class
* AddCommand Class
* DeleteCommand Class
* MarkCommand Class
* UnmarkCommand Class
* ListCommand Class

nusE0726844 committed:
Tweak the code to comply with coding standard

nusE0726844 committed:
Add FindCommand()

Give users a way to find a task by searching for a keyword

nusE0726844 committed:
Merge branch 'branch-A-CodingStandard'

nusE0726844 committed:
Merge branch 'branch-Level-9'

nusE0726844 committed:
Add checkstyle
Use checkstyle to detect coding style violations

nusE0726844 committed:
Add GUI
Use JavaFX to implement GUI

nusE0726844 committed:
Merge branch 'branch-Level-10'

nusE0726844 committed:
Use var-args

nusE0726844 committed:
Merge branch 'branch-A-Varargs'

nusE0726844 committed:
Rewrite command class methods

Command class methods have void return types. These methods have
in-built methods to show responses for text ui.

Rewriting these methods to return response allows responses to be shown
on gui ui.

nusE0726844 committed:
Resolve checkstyle

nusE0726844 committed:
Add assertions

nusE0726844 committed:
Add stream

findCommand() uses a for loop to iterate over an arraylist to check for keyword.

Using stream to filter the tasklist for the keyword makes it easier to maintain.

nusE0726844 committed:
Improve code quality

nusE0726844 committed:
Merge pull request #1 from nusE0726844/branch-A-Assertions

[Kelvin Fong Koi Mun] iP A-Assertions

nusE0726844 committed:
Merge branch 'branch-A-CodeQuality'

nusE0726844 committed:
Merge branch 'branch-A-Streams'

nusE0726844 committed:
Add gradle.yml

Use GitHub Actions to set up Continuous Integration (CI).

nusE0726844 committed:
Add C-DetectDuplicates

Add the ability to recognize duplicate tasks being adding to TaskList.

nusE0726844 committed:
Refactor Task classes for SLAP

nusE0726844 committed:
Refactor Command classes for SLAP

nusE0726844 committed:
Move TaskList class to task package

Fixed javadoc for Command classes.

nusE0726844 committed:
Improve the GUI

Added features:
* Highlight errors
* Tweak padding
* Tweak colors

nusE0726844 committed:
Refactor DukeException to handle errors

nusE0726844 committed:
Add more JUnit tests

More JUnit tests were added so that nearly all code that can be tested automatically.

nusE0726844 committed:
Choose a unique personality for Duke

Gave Duke a personality named Jamie. Tweaked the welcome message to include an introduction.

nusE0726844 committed:
Merge branch 'branch-A-Personality'

nusE0726844 committed:
Merge branch 'branch-A-MoreErrorHandling'

# Conflicts:
#	src/main/java/duke/Duke.java

nusE0726844 committed:
Merge branch 'branch-A-MoreTesting'

nusE0726844 committed:
Change JUnit tests

nusE0726844 committed:
Tweaked GUI

nusE0726844 committed:
Add a User Guide

nusE0726844 committed:
Tweaked GUI to include colored error messages

nusE0726844 committed:
Reformat user guide

nusE0726844 committed:
Reformat user guide
